standard/attributes.md

```diff
@@ -858,6 +860,275 @@ For invocations that occur within field or event initializers, the member name u
+`NotNullIfNotNull` (§The-NotNullIfNotNull-Attribute)  | A return value isn't null if the argument for the specified parameter isn't null.
+`MemberNotNull` (§The-MemberNotNull-Attribute)  | The listed member won't be null when the method returns.
+`MemberNotNullWhen` (§The-MemberNotNullWhen-Attribute)  | The listed member won't be null when the method returns the specified `bool` value.
+`DoesNotReturn` (§The-DoesNotReturn-Attribute)  | This method never returns.
+`DoesNotReturnIf` (§The-DoesNotReturnIf-Attribute)  | This method never returns if the associated `bool` parameter has the specified value.
+
+The following sections in §Nullable-Analysis-Attributes-General are conditionally normative.
+
+#### §The-AllowNull-Attribute The AllowNull attribute
+
+Specifies that a null value is allowed as an input even if the corresponding type disallows it.
+
+> *Example*: Consider the following read/write property that never returns `null` because it has a reasonable default value. However, a user can give null to the set accessor to set the property to that default value.
+>
+> <!-- Example: {template:"standalone-lib", name:"AllowNullAttribute", replaceEllipsis:true, customEllipsisReplacements:["\"XYZ\""]} -->
+> ```csharp
+> #nullable enable
+> public class X
+> {
+>     [AllowNull]
+>     public string ScreenName
+>     {
+>         get => _screenName;
+>         set => _screenName = value ?? GenerateRandomScreenName();
+>     }
+>     private string _screenName = GenerateRandomScreenName();
+>     private static string GenerateRandomScreenName() => ...;
+> }
+> ```
+>
+> Given the following use of that property’s set accessor
+>
+> ```csharp
+> var v = new X();
+> v.ScreenName = null;   // may warn without attribute AllowNull
+> ```
+>
+> without the attribute, the compiler may generate a warning because the non-nullable-typed property appears to be set to a null value. The presence of the attribute suppresses that warning. *end example*
+
+#### §The-DisallowNull-Attribute The DisallowNull attribute
+
+Specifies that a null value is disallowed as an input even if the corresponding type allows it.
+
+> *Example*: Consider the following property in which null is the default value, but clients can only set it to a non-null value.
+>
+> <!-- Example: {template:"standalone-lib", name:"DisallowNullAttribute"} -->
+> ```csharp
+> #nullable enable
+> public class X
+> {
+>     [DisallowNull]
+>     public string? ReviewComment
+>     {
+>         get => _comment;
+>         set => _comment = value ?? throw new ArgumentNullException(nameof(value),
+>            "Cannot set to null");
+>     }
+>     private string? _comment = default;
+> }
+> ```
+>
+> The get accessor could return the default value of `null`, so the compiler may warn that it must be checked before access. Furthermore, it warns callers that, even though it could be null, callers shouldn't explicitly set it to null. *end example*
+
+#### §The-DoesNotReturn-Attribute The DoesNotReturn attribute
```

Comment:
This feels only tangentially null-related. Maybe we should specify this one normatively? (Ditto DoesNotReturnIf)

Reply:
I'll ask compiler folks, but I don't think these attributes affect any other analysis (such as definite assignment, or unreachable code).

Reply:
I updated the sample to provide one that more likely matches a typical use. It validates an argument as to whether it's null or not.

Reply:
Righto.